File: hooks/writing-suggest-verify.py
#!/usr/bin/env -S uv run python3
"""
PostToolUse hook: Suggest /writing-revise after N edits in writing workflow.

Only fires for Edit/Write on .md files when an active writing workflow exists.
Tracks edit count in ACTIVE_WORKFLOW.md and suggests edit loop at threshold.
"""
import json
import re
import sys
from pathlib import Path

# Hooks read their payload from STDIN -- CLAUDE_TOOL_INPUT does not exist -- and
# {"result": "continue"} is not part of the hook contract. This hook used both, so it
# saw an empty file_path on EVERY Edit/Write, took the early-return branch, and then
# emitted a payload the harness rejected outright ("Hook JSON output validation
# failed"). Net effect: the verify-nudge never fired and the edit counter never moved.
# Non-blocking feedback on PostToolUse is hookSpecificOutput.additionalContext;
# printing nothing is how a hook says "carry on".
HOOKS_DIR = Path(__file__).resolve().parent
sys.path.insert(0, str(HOOKS_DIR))
from _gate_common import context  # noqa: E402


def parse_yaml_value(content: str, key: str, default=None):
    """Extract a single value from YAML-like content."""
    pattern = rf'^{key}:\s*(.+)$'
    match = re.search(pattern, content, re.MULTILINE)
    if match:
        return match.group(1).strip().strip('"').strip("'")
    return default


def update_yaml_value(content: str, key: str, new_value) -> str:
    """Update a single value in YAML-like content."""
    pattern = rf'^({key}:\s*)(.+)$'
    replacement = rf'\g<1>{new_value}'
    new_content, count = re.subn(pattern, replacement, content, flags=re.MULTILINE)
    if count == 0:
        # Key doesn't exist, add it after the opening ---
        lines = content.split('\n')
        for i, line in enumerate(lines):
            if line.strip() == '---' and i > 0:
                # Found the closing ---, insert before it
                lines.insert(i, f'{key}: {new_value}')
                return '\n'.join(lines)
        # No closing --- found, append to end
        return content + f'\n{key}: {new_value}'
    return new_content


def main():
    try:
        hook_input = json.load(sys.stdin)
    except Exception:
        sys.exit(0)

    if hook_input.get('tool_name', '') not in ('Write', 'Edit', 'MultiEdit'):
        sys.exit(0)

    tool_input = hook_input.get('tool_input', {}) or {}
    file_path = tool_input.get('file_path', '')

    # Only process markdown files
    if not file_path.endswith('.md'):
        sys.exit(0)

    # Check for active writing workflow
    workflow_path = Path.cwd() / '.planning' / 'ACTIVE_WORKFLOW.md'
    if not workflow_path.exists():
        sys.exit(0)

    try:
        content = workflow_path.read_text()
    except Exception:
        sys.exit(0)

    # Check if this is a writing workflow
    workflow_type = parse_yaml_value(content, 'workflow')
    if workflow_type != 'writing':
        sys.exit(0)

    # Get current edit count and threshold
    edits = int(parse_yaml_value(content, 'edits_since_verify', '0'))
    threshold = int(parse_yaml_value(content, 'verify_threshold', '10'))

    # Increment edit count
    edits += 1

    if edits >= threshold:
        # Reset counter and suggest verification
        new_content = update_yaml_value(content, 'edits_since_verify', '0')
        workflow_path.write_text(new_content)

        style = parse_yaml_value(content, 'style', 'general')
        phase = parse_yaml_value(content, 'phase', 'edit')

        context(
            'PostToolUse',
            f"📝 {edits} edits since last verify (style: {style}, phase: {phase}). "
            f"Consider `/writing-revise` to apply fixes and polish."
        )
    else:
        # Just increment counter
        new_content = update_yaml_value(content, 'edits_since_verify', str(edits))
        workflow_path.write_text(new_content)
        sys.exit(0)


if __name__ == '__main__':
    main()

File: hooks/writing-suggest-verify.ts
#!/usr/bin/env bun
/** PostToolUse hook: suggest /writing-revise after a plan-bound edit threshold.
 *
 * The active writing identity is the APPROVED receipt-selected generated plan.
 *
 * THE COUNTER MOVED. It was `.planning/.state/writing.json` — a per-workflow state file with exactly
 * one consumer, this hook. It now lives in `editsSinceVerify` on the shared
 * `.planning/.state/episode.json`, because one consumer is not a reason for a file and that pattern
 * is how the planning directory reached eight state files across three classes. See
 * `.claude/CLAUDE.md` -> "State Files". Retired `ACTIVE_WORKFLOW.md` is conversion input only and is
 * never read or written.
 */
import { existsSync } from "node:fs";
import { context, readPayload } from "./_gate_common.ts";
import { authenticatedWritingPlan } from "./lib/writing-plan-context.ts";
import { episodeStatePath, initEpisodeState, matchesPlan, readEpisodeState, writeEpisodeState, type EpisodeState } from "./lib/episode-state.ts";

const DEFAULT_THRESHOLD = 10;

/**
 * The episode this edit belongs to, or `null` when the hook must not act.
 *
 * A MISSING state is created — an approved plan is authenticated by this point, so there is a real
 * episode to count against. A state bound to a DIFFERENT plan is not created over: the plan moved,
 * and resetting another episode's record to count edits would discard a recorded phase or a review
 * debt. That case returns null and the hook stays silent, which is the same fail-inert posture the
 * old `loadState` took on a malformed file.
 */
function episodeFor(projectRoot: string, workflow: string, planFile: string, planHash: string, sessionId: string | null): EpisodeState | null {
  const existing = readEpisodeState(projectRoot);
  if (existing === null) {
    // ABSENT AND UNREADABLE BOTH COME BACK AS `null`, AND CONFLATING THEM DESTROYS STATE.
    // Creating a fresh episode over a file we could not parse overwrites recorded phases and any
    // outstanding review debt — silently discharging a review nobody performed. The paragraph above
    // has always claimed this guard; for one evening it did not implement it, while the identical
    // guard was written correctly in `episode-phase.ts` and `implement-gate.ts`. Found by the gemini
    // third-party adapter reviewing this very change; codex reviewed the same diff and missed it.
    if (existsSync(episodeStatePath(projectRoot))) return null;
    return initEpisodeState({ workflow, sessionId, planFile, planHash });
  }
  if (!matchesPlan(existing, planFile, planHash)) return null;
  // A pre-plan episode gets its binding filled in the first time an approved plan is seen, and a
  // null session is backfilled the first time a writer HAS one. Leaving sessionId null forever meant
  // `clarify-before-recon-guard` could never match it and silently fell back to the sentinel — the
  // trustworthy evidence path quietly switching itself off. Found by the gemini adapter.
  return {
    ...existing,
    ...(existing.planHash === null ? { planFile, planHash } : {}),
    ...(existing.sessionId === null && sessionId !== null ? { sessionId } : {}),
  };
}

async function main(): Promise<void> {
  let payload: Record<string, unknown>;
  try {
    payload = await readPayload();
  } catch {
    return;
  }
  const toolName = payload.tool_name;
  if (toolName !== "Write" && toolName !== "Edit" && toolName !== "MultiEdit") return;
  const filePath = (payload.tool_input as Record<string, unknown> | undefined)?.file_path;
  if (typeof filePath !== "string" || !filePath.endsWith(".md")) return;

  const plan = authenticatedWritingPlan(filePath);
  if (!plan) return;
  // No session in this payload is normal, not a failure — the episode records it as null and a
  // later session-bound writer fills it. Refusing here would silently stop counting edits.
  const sessionId = typeof payload.session_id === "string" && payload.session_id.trim() ? payload.session_id : null;
  // A malformed, tampered, or stale state must not be overwritten by a hook.
  const state = episodeFor(plan.projectRoot, "writing", plan.plan.planFile, plan.plan.hash, sessionId);
  if (!state) return;

  const edits = state.editsSinceVerify + 1;
  if (edits < DEFAULT_THRESHOLD) {
    writeEpisodeState(plan.projectRoot, { ...state, editsSinceVerify: edits });
    return;
  }
  writeEpisodeState(plan.projectRoot, { ...state, editsSinceVerify: 0 });
  context(
    "PostToolUse",
    `📝 ${edits} edits since last verify (style: ${plan.style || "general"}, phase: edit). ` +
      "Consider `/writing-revise` to apply fixes and polish.",
  );
}

await main();
